storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,669 @@
+        Map<String, Integer> approvedWorkers = localState.getApprovedWorkers();
+        Set<String> approvedIds = new HashSet<>();
+        if (approvedWorkers != null) {
+            approvedIds.addAll(approvedWorkers.keySet());
+        }
+        for (Map.Entry<String, LSWorkerHeartbeat> entry : idToHeartbeat.entrySet()) {
+            String workerId = entry.getKey();
+            LSWorkerHeartbeat whb = entry.getValue();
+            State state;
+            if (whb == null) {
+                state = State.notStarted;
+            } else if (!approvedIds.contains(workerId) || !matchesAssignment(whb, assignedExecutors)) {
+                state = State.disallowed;
+            } else if (supervisorData.getDeadWorkers().contains(workerId)) {
+                LOG.info("Worker Process {}as died", workerId);
+                state = State.timedOut;
+            } else if (SupervisorUtils.isWorkerHbTimedOut(now, whb, conf)) {
+                state = State.timedOut;
+            } else {
+                state = State.valid;
+            }
+            LOG.debug("Worker:{} state:{} WorkerHeartbeat:{} at supervisor time-secs {}", workerId, state, whb, now);
+            workerIdHbstate.put(workerId, new StateHeartbeat(state, whb));
+        }
+        return workerIdHbstate;
+    }
+
+    protected boolean matchesAssignment(LSWorkerHeartbeat whb, Map<Integer, LocalAssignment> assignedExecutors) {
+        LocalAssignment localAssignment = assignedExecutors.get(whb.get_port());
+        if (localAssignment == null || !localAssignment.get_topology_id().equals(whb.get_topology_id())) {
+            return false;
+        }
+        List<ExecutorInfo> executorInfos = new ArrayList<>();
+        executorInfos.addAll(whb.get_executors());
+        // remove SYSTEM_EXECUTOR_ID
+        executorInfos.remove(new ExecutorInfo(-1, -1));
+        List<ExecutorInfo> localExecuorInfos = localAssignment.get_executors();
+
+        if (localExecuorInfos.size() != executorInfos.size())
+            return false;
+
+        for (ExecutorInfo executorInfo : localExecuorInfos){
+            if (!localExecuorInfos.contains(executorInfo))
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * launch a worker in local mode.
+     */
+    protected void launchWorker(SupervisorData supervisorData, String stormId, Long port, String workerId, WorkerResources resources) throws IOException {
+        // port this function after porting worker to java
+    }
+
+    protected String getWorkerClassPath(String stormJar, Map stormConf) {
+        List<String> topoClasspath = new ArrayList<>();
+        Object object = stormConf.get(Config.TOPOLOGY_CLASSPATH);
+
+        if (object instanceof List) {
+            topoClasspath.addAll((List<String>) object);
+        } else if (object instanceof String){
+            topoClasspath.add((String)object);
+        }else {
+            //ignore
```

Comment:
The original code was

```
(let [...
          topo-classpath (if-let [cp (storm-conf TOPOLOGY-CLASSPATH)]
                           [cp]
                           [])
          ... (Utils/addToClasspath something topo-classpath) ...
```

The Signature for Utils.addToClasspath is

```
public static String addToClasspath(String classpath, Collection<String> paths);
```

if `cp` was anything but a `String` in that code we would get a `ClassCastException` that would bring down the supervisor.  This code is even more lenient and supports both `String` and `List<String>`.  Which fixes a bug in our translation of the Utils code.

You are right that we don't want a bad user config to bring down a supervisor, but if nimbus is doing the right thing and validating the config before it goes to the supervisor we should never see this happen.  `TOPOLOGY_CLASSPATH` is annotated with ` @isStringOrStringList`.  So if we ever see `object` as something that is not a `String`, `List<String>`, or `null` it means we have a bug in our code somewhere, and silently ignoring it feels wrong to me.

I would be fine with us logging an error if it is something different, but not totally ignoring it.